### PR TITLE
Record access and assignment

### PR DESCRIPTION
**Pull request by buggymcbugfix**
Named records are clumsy to use without record access and assignment. E.g. consider the following types and global variable:

```c#
type Address = Addr(text street, text city);
type Person = Pers(text name, Address address);

Person bob = Pers("bob", Addr("main st", "springfield"));
```

We might want to change the street where bob lives. Currently we have to do this:
```c#
  bob = case bob {
    Pers(name, address) -> case address {
      Addr(street, city) -> Pers(name, Addr(newStreet, city));
    }
  }
```

With assignment to record fields, we can simplify this to:
```c#
  bob.address.street = newStreet
```

Similarly nested record access will make code simpler.

This could be implemented via desugaring but this would make it hard to report errors accurately. Therefore I implement it as  a primitive in the language.

Status (checked -> done, unchecked -> in progress)

- [x] determine what constitutes a valid field when there is more than one constructor
- [x] nested record access expressions
  - [x] parse
  - [x] typecheck
  - [x] eval
  - [x] undefinedness analysis
  - [x] tests
- [x] nested record assignments (orthogonal to expressions)
  - [x] parse
  - [x] typecheck
  - [x] eval
  - [x] undefinedness analysis
  - [x] tests

### CODE REVIEW

**Comment by buggymcbugfix** [comment]
TODO:
- [ ] merge master
- [ ] update Uplink
- [ ] update documentation

**Comment by buggymcbugfix** [comment]
I am now doubting my choice to use upcased identifiers for user defined types, e.g. `type T = C`. If we add generics (parametric polymorphism) in future, it seems more consistent with other languages (except Haskell) to use uppercase for the type variables, e.g.:
```c#
type list<A> = Nil | Cons(A head, list<A> tail);
```

**Comment by sdiehl** [comment]
@buggymcbugfix Update the Changelog.

**Comment by sdiehl** [comment]
This will need to be updated to track master.

File: src/Language/FCL/Parser.hs [hunk 1]
```diff
@@ -664,21 +670,27 @@ transition = do
   <?> "transition"
 
 -------------------------------------------------------------------------------
--- ADTeration type definition
+-- Algebraic datat type definition
```
Comment (Acentelles) [inline src/Language/FCL/Parser.hs]: "datat"

Comment (buggymcbugfix) [reply src/Language/FCL/Parser.hs]: Thanks

**Comment by buggymcbugfix** [comment]
This branch should be ready to merge. Next step is documentation and updating `uplink`.

**Comment by sdiehl** [comment]
@Acentelles Can you give this a look over and see if it's aligned with your staged changes?

**Comment by buggymcbugfix** [comment]
Please don't merge this yet, @sdiehl, I'm going to tweak something.

File: src/Language/FCL/AST.hs [hunk 2]
```diff
@@ -276,13 +276,13 @@ data Expr
   | EBefore  LExpr LExpr           -- ^ Time guard
   | EAfter   LExpr LExpr           -- ^ Time guard
   | EBetween LExpr LExpr LExpr     -- ^ Time guard
-  | ECase    LExpr [CaseBranch]         -- ^ Case statement
-  | EAssign  Name  LExpr           -- ^ Variable update
+  | ECase    LExpr [CaseBranch]    -- ^ Case statement
+  | EAssign  (NonEmpty Name)  LExpr -- ^ Variable update
```
Comment (sdiehl) [inline src/Language/FCL/AST.hs]: Nice.

File: src/Language/FCL/AST.hs [hunk 3]
```diff
@@ -670,42 +671,14 @@ mapType einfo   (VSet vset)   =
...
-  { constructorToType :: Map NameUpper (LNameUpper, [(Type, LName)])
-  , adtToConstrs :: Map NameUpper [ADTConstr]
+  { constructorToType :: Map NameUpper (LName, [(LName, Type)])
+  , adtToConstrsAndFields :: Map Name (NonEmpty ADTConstr, [(Name, Type)])
   }
 
--- | Create the dictionaries for the constructor/adt type membership
--- relations from the original list of definitionSet. Assumes that there
--- are no duplicates in the input.
-createADTInfo :: [ADTDef] -> ADTInfo
```
Comment (Acentelles) [inline src/Language/FCL/AST.hs]: @buggymcbugfix why was this function removed?

**Comment by buggymcbugfix** [comment]
I believe that I moved it to `Typecheck.hs`. Did I not export it from there?

**Pull request merged**